snap/gadget.go

```diff
@@ -69,6 +72,71 @@ type VolumeContent struct {
 	Unpack bool `yaml:"unpack"`
 }
 
+// GadgetConnect describes an interface connection requested by the gadget
+// between seeded snaps. The syntax looks like this:
+//
+// [<plug-snap-id1>|system]:plug [[<slot-snap-id>|system]:slot]
+//
+// "system" or omitting the snap-id indicates a system plug or slot.
+// Fully omitting the slot part indicates a system slot with the same name
+// as the plug.
+type GadgetConnect struct {
+	PlugSnapID string
+	Plug       string
+	SlotSnapID string
+	Slot       string
+}
+
+func parseSnapIDColonName(s string) (snapID, name string, err error) {
+	parts := strings.Split(s, ":")
+	if len(parts) == 2 {
+		snapID = parts[0]
+		name = parts[1]
+	}
+	if name == "" {
+		return "", "", fmt.Errorf("in gadget connect expected [snap-id]:name not %q", s)
+	}
+	if snapID == "system" {
+		snapID = ""
+	}
+	return snapID, name, nil
+}
+
+func (gconn *GadgetConnect) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var s string
+	if err := unmarshal(&s); err != nil {
+		return err
+	}
+	flds := strings.Fields(s)
+	switch len(flds) {
+	case 2:
+		plugSnapID, plug, err := parseSnapIDColonName(flds[0])
+		if err != nil {
+			return err
+		}
+		slotSnapID, slot, err := parseSnapIDColonName(flds[1])
+		if err != nil {
+			return err
+		}
+		gconn.PlugSnapID = plugSnapID
+		gconn.Plug = plug
+		gconn.SlotSnapID = slotSnapID
+		gconn.Slot = slot
+	case 1:
+		plugSnapID, plug, err := parseSnapIDColonName(flds[0])
+		if err != nil {
+			return err
+		}
+		gconn.PlugSnapID = plugSnapID
+		gconn.Plug = plug
+		gconn.SlotSnapID = ""
+		gconn.Slot = plug
```

Comment:
Nitpick: you could probably make it much shorter and avoid some repetition by having something like:
```var plugSnapID, plug, slotSnapID, slot string
 switch(len(flds)) {
   case 2: 
      slotSnapID, slot, err = parseSnapIDColonName(flds[1])
      if err { .... }
      fallthrough
  case 1:
     plugSnapID, plug, err = parseSnapIDColonName(flds[0])
     if err { ... }
 default: ....
 }
gconn.PlugSnapID = plugSnapID
...
gconn.SlotSnapID = slotSnapID
```
But I'm not going to block on this.

Reply:
I thought of that, it's a bit annoying because in the case 1 we also need  gconn.Slot = plug

Reply:
Right, I missed that. It's fine then.